docs/how-tos/build-opa-bundle.md

```diff
@@ -0,0 +1,123 @@
+# Defining and building policy bundles
+
+## Preface
+
+This guide will explain how to write and build rego language policy into a bundle which can be used as an entrypoint to your AuthZ decisions, the "Application Policy" referred to in the preface of the [configuring OPA how-to](configure-opa.md).
+
+See also [the OPA documentation on Rego](https://www.openpolicyagent.org/docs/latest/policy-language/) and the [opa playground](https://play.openpolicyagent.org/).
+
+## Writing policy
+
+Below is an example of rego policy that uses a dependent package named `diamond` which defines granular authorization rules: it is assumed that the dependent package defines two methods: `user_on_session(proposalNumber, visitNumber)` and `user_on_proposal(proposalNumber)`. While OPA typically expects all policy to be defined within a single bundle, in an attempt to abstract the moving target of Diamond Authentication and to simplify adoption, these functions will be defined in `org-policy` in this repository and made available to write policy against.
+
+In this example the policy is being used for an example service which serves data over HTTP:
+
+- The path <service.diamond.ac.uk/$proposalNumber> lists a number of sessions, which a user should only be able to see if they are on the proposal.  
+- The path <service.diamond.ac.uk/$proposalNumber/$visitNumber> lists a collection of data files, which the user should only be able to see if they are on the session.  
+- The path <service.diamond.ac.uk/$proposalNumber/$visitNumber/$fileName> downloads a file from the service, if the user is on its session.
+
+It is assumed that the user has already been authenticated, and user authentication information is contained within the headers of the request. The exact form of the authentication information, and how to parse it is left to the `diamond` package to define. The `input` to the OPA instance would include the following structure:
```

Comment:
Will `input` include this structure for the python example, or only when using itsio?

Reply:
The `input` ought to look pretty much identical between the two methods, I will check but the fastapi middleware should basically forward the whole request sans body, same as Istio

Reply:
While awaiting resolution of [this issue](https://github.com/busykoala/fastapi-opa/issues/68) on fastapi-opa middleware (which seems the easiest, most complete solution for deploying Python app with OPA AuthZ), I've tested each individual part of the plugin but not the overall flow.

It looks like the exact structure of the input is implementation specific: Istio uses "parsed_path", fastapi-opa uses "request_path", the example on the opa docs uses "path". 

Luckily these kinds of implementation detail will have to be left to the application specific policy, but that's OK. The org-policy in this bundle just exposes a series of methods as endpoints, applications write their own policy that can call into the policy to make use of the bundled permissionables if they want.

Reply:
The overall structure of being a json blob will be the same though